### PR TITLE
Add support for Cardano legacy's encrypted secret keys. 

### LINKED ISSUE #5186: Add support for Cardano legacy's encrypted secret keys.
### Add support for Cardano legacy's encrypted secret keys.

In 2017 the old wallets used encrypted master keys for keeping the users funds safe.
Unfortunately, lot of users lost their mnemonic and forgot their spending/encryption password, that caused locking their funds forever.

The PR I created tackles this issues by trying to decrypt the encrypted keys.
Unfortunately, it requires a slight modification in the ed25519-donna package as it does support extended (64-byte length) secret keys. This minor modification adds the possibility of generating a ed25519 public key from an extended secret key.

The Cardano legacy's extended secret key is 128-byte length that contains: a
1. 64-byte length encrypted secret key (from entropy), a
2. 32-byte length ed25519 public key (not-encrypted) and a
3. 32-byte length chaincode from entropy (not-encrypted)

To decrypt an encrypted (master or any derived) key needs the following steps:
- `blake2b_256` hash of the password candidate,
- a 40-byte long key from `pbkdf2_hmac_sha512`, using a static salt, and the pasword hash,
- decrypt the encrypted secret using `ChaCha20` stream encryption alg with the derived key (as IV and Key) and
- generate a public from the decrypted candidate secret key and 
- compare it with the store public key. If the both match then the password is found.

Cardano had a second method to validate whether the secret key was encrypted or not, by using `Scrypt` with some random salt.
Though, using the `scrypt`, to find the encryption password, is much slower than the above and also it does not guarantee (due to some bug in the early wallet implementations) that password generated the scrypt based password hash can decrypt the encrypted key at all.

### PR DESCRIPTION
**Pull request by ilap**
Resolves #5186

### CODE REVIEW

**Review by solardiz — CHANGES_REQUESTED** [review]
Thank you for the contribution, @ilap! Let's polish it a bit and then likely get it merged.

To ease review, what format's source files did you base yours on?

There are a few issues here right now:

In my testing, this fails self-test when run with more than one OpenMP thread. The failures are inconsistent - they're `FAILED (cmp_all(14))`, but with indices changing (`14` in this example, but it varies). This suggests there's a race condition or/and usage of uninitialized memory.

(The good news, though, is the `tezos` format still works fine for me after your changes to the shared code.)

There's only one test vector. Please add at least one more, for 2+ total. Some realistic issues will go undetected when there's only one test vector.

You do split out a few things into `common` files, but you don't actually have an OpenCL format yet - do you intend to contribute one with your next PR? If not, then at least comments need to be edited to say these files are `common` for their future use, and currently are not, and there isn't currently an OpenCL format. BTW, we do have all of the required primitives in OpenCL, so it shouldn't be too hard to add an OpenCL format implementing this.

Thank you for including our preferred copyright+license statement in the main `cardano_fmt_plug.c` file. Please also add that to other source files you're adding here.

The `ALGORITHM_NAME` is wrong. Please change it perhaps to:

```
#ifdef SIMD_COEF_64
#define ALGORITHM_NAME      "PBKDF2-SHA512/BLAKE2b/ChaCha20 " SHA512_ALGORITHM_NAME
#else
#if ARCH_BITS >= 64
#define ALGORITHM_NAME      "PBKDF2-SHA512/BLAKE2b/ChaCha20 64/" ARCH_BITS_STR
#else
#define ALGORITHM_NAME      "PBKDF2-SHA512/BLAKE2b/ChaCha20 32/" ARCH_BITS_STR
#endif
```

(The most time-consuming component goes first, and the SIMD'ness and bitness at the end corresponds to its implementation's use of CPU.)

I and/or others will likely have more feedback later, perhaps after you address the above.

Thanks again!

**Comment by solardiz** [comment]
Somehow Circle CI doesn't run on this PR - given that it ran just fine on a nearby PR by someone else, perhaps their abuse detection somehow got (wrongly) triggered for @ilap? This happened for some of our contributors IIRC a year ago, but didn't happen since...

**Comment by solardiz** [comment]
> this fails self-test when run with more than one OpenMP thread.

The below fixes the obvious issues, and makes tests pass for me (but we need more test vectors to be more certain, we're lucky the issue was detected at all with only one test vector):

```diff
+++ b/src/cardano_fmt_plug.c
@@ -125,8 +125,8 @@ static int crypt_all(int *pcount, struct db_salt *salt)
        for (index = 0; index < count; index += MIN_KEYS_PER_CRYPT) {
                int i;
 
-               static unsigned char kdf_out[MIN_KEYS_PER_CRYPT][BUF_SIZE];
-               static unsigned char blake_out[MIN_KEYS_PER_CRYPT][PWD_HASH_LEN];
+               unsigned char kdf_out[MIN_KEYS_PER_CRYPT][BUF_SIZE];
+               unsigned char blake_out[MIN_KEYS_PER_CRYPT][PWD_HASH_LEN];
 
 #if SIMD_COEF_64
                int lens[MIN_KEYS_PER_CRYPT];
@@ -151,12 +151,12 @@ static int crypt_all(int *pcount, struct db_salt *salt)
                }
 #endif
 
-               static ed25519ext_secret_key dsks[MIN_KEYS_PER_CRYPT];
-               static ed25519_public_key pk;
+               ed25519ext_secret_key dsks[MIN_KEYS_PER_CRYPT];
+               ed25519_public_key pk;
 
                for (i = 0; i < MIN_KEYS_PER_CRYPT; ++i) {
 
-                       static struct chacha_ctx ckey;
+                       struct chacha_ctx ckey;
 
                        // Decrypt the encrypted esk with the retrieved KDFs.
                        chacha_keysetup(&ckey, kdf_out[i], KEY_SIZE * 8);
```

**Comment by solardiz** [comment]
Oh, also please avoid over-long commit message lines. Please split lines:

```
Add support for Cardano legacy's encrypted secret keys

Resolves #5186
```

**Comment by solardiz** [comment]
> Somehow Circle CI doesn't run on this PR

Logged in to Circle CI, and I don't see this PR's jobs there at all. There are for other recent activities. Also, we appear to be very far from bumping into any resource limit this month.

Maybe just a glitch. Maybe they will pick this up on the next update to this PR.

**Comment by ilap** [comment]
Sorry, different timezone.

> To ease review, what format's source files did you base yours on?

It's a mix and match, just had a look at some source codes and got hints from them to implement.

> You do split out a few things into `common` files, but you don't actually have an OpenCL format yet - do you intend to contribute one with your next PR? If not, then at least comments need to be edited to say these files are `common` for their future use, and

I think it will be `common` for a little while, but I might will create some `cl` code and some python tools for easier usability.

> (The good news, though, is the `tezos` format still works fine for me after your changes to the shared code.)

Yeah, it just adds an extra functionality, no breaking API changes.


> There's only one test vector. Please add at least one more, for 2+ total. Some realistic issues will go undetected when there's only one test vector.

Unfortunately, I could get only one proper testvector (encrypted and has its en/decryption key) by retrieving it from some old keystore file (`secret.key`) that was/is used for testing some old functionality of the officiall wallet.

> Oh, also please avoid over-long commit message lines. Please split lines:

Will do, thank you.

> Thank you for including our preferred copyright+license statement in the main `cardano_fmt_plug.c` file. Please also add that to other source files you're adding here.

Done

> (The most time-consuming component goes first, and the SIMD'ness and bitness at the end corresponds to its implementation's use of CPU.)

Done 

> The below fixes the obvious issues, and makes tests pass for me (but we need more test vectors to be more certain, we're lucky the issue was detected at all with only one test vector):

Thank you, I have removed the `static` keywords and it makes sense why it's failing.

I will commit the changes soon, thx again.

**Comment by ghost** [comment]
Since no CicleCI job is running again, I'm running the upstream CI at https://app.circleci.com/pipelines/github/claudioandre-br/JohnTheRipper/478/workflows/1292c785-5593-44c9-9fb6-e6a6c2b1f04e

The patch adds trailing spaces. 

```
warning: 2 lines add whitespace errors.
```

**Comment by ilap** [comment]
> Since no CicleCI job is running again, I'm running the upstream CI at https://app.circleci.com/pipelines/github/claudioandre-br/JohnTheRipper/478/workflows/1292c785-5593-44c9-9fb6-e6a6c2b1f04e
> 
> The patch adds trailing spaces.
> 
> ```
> warning: 2 lines add whitespace errors.
> ```

What tools do you use to check  whitespace errors and which parameters?
As my one does not recognised trailing spaces in comments.

**Comment by ilap** [comment]
> Since no CicleCI job is running again, I'm running the upstream CI at https://app.circleci.com/pipelines/github/claudioandre-br/JohnTheRipper/478/workflows/1292c785-5593-44c9-9fb6-e6a6c2b1f04e
> 
> The patch adds trailing spaces.
> 
> ```
> warning: 2 lines add whitespace errors.
> ```

Fixed

**Comment by solardiz** [comment]
> some python tools for easier usability.

Yes, and/or please add some documentation on how to use this. (Not necessarily in this same PR.)

> Unfortunately, I could get only one proper testvector (encrypted and has its en/decryption key) by retrieving it from some old keystore file (`secret.key`) that was/is used for testing some old functionality of the officiall wallet.

OK, so we'll have to accept having only one for now.

Can you please add the `secret.key` file and its corresponding password via a pull request to https://github.com/openwall/john-samples?

Please squash all commits here into one, and if you make any further changes please amend and force-push the one commit instead of adding fixup commits. (Alternatively, we can use GitHub's squash-and-merge, but that makes GitHub the committer per git history and it's a feature we normally keep disabled for this repo, to avoid inadvertent use.)

> Since no CicleCI job is running again, I'm running the upstream CI at https://app.circleci.com/pipelines/github/claudioandre-br/JohnTheRipper/478/workflows/1292c785-5593-44c9-9fb6-e6a6c2b1f04e

Thank you, @claudioandre-br! As I can see, this passed all of our tests now. Great.

**Comment by ilap** [comment]
> Can you please add the `secret.key` file and its corresponding password via a pull request to https://github.com/openwall/john-samples?

Uploaded and a PR is created.

> Please squash all commits here into one, and if you make any further changes please amend and force-push the one commit instead of adding fixup commits. (Alternatively, we can use GitHub's squash-and-merge, but that makes GitHub the committer per git history and it's a feature we normally keep disabled for this repo, to avoid inadvertent use.)

Done

File: src/cardano_common.h
Comment (solardiz) [inline src/cardano_common.h]: Out of these 3, probably only `#include "formats.h"` is needed here.

File: src/cardano_common.h
Comment (solardiz) [inline src/cardano_common.h]: Should be "This file" (singular), and ditto in the other file with that comment.

File: src/cardano_common.h
Comment (solardiz) [inline src/cardano_common.h]: There isn't actually a `cardano_iteration_count()` - so drop the prototype.

File: src/cardano_common_plug.c
Comment (solardiz) [inline src/cardano_common_plug.c:60]: Doesn't this allow any trailing garbage after the hash? Perhaps it should insist on no trailing garbage?

File: src/cardano_common_plug.c
Comment (solardiz) [inline src/cardano_common_plug.c:52]: If we currently only support `1` there, we should insist on that - so e.g. instead of `!isdec(p)`, we can `strcmp(p, "1")`.

Also, we don't have to decode that field into the salt structure until we use it, nor to have a field in there. We can just have a comment on its possible future use, here in `valid`.

File: src/cardano_fmt_plug.c
Comment (solardiz) [inline src/cardano_fmt_plug.c:30]: A few of these `#include` are probably irrelevant, I guess at least `#include "options.h"`.

File: src/cardano_fmt_plug.c
Comment (solardiz) [inline src/cardano_fmt_plug.c:55]: I think 15000 iterations is so high that the `* 4` and a non-SIMD max of 4 are unneeded - can set them to 1 for greater program interactivity at no significant effect on performance.

File: src/cardano_fmt_plug.c
Comment (solardiz) [inline src/cardano_fmt_plug.c:59]: Why limit `PLAINTEXT_LENGTH` to 15 here? Looks wrong to me, perhaps copy-paste from elsewhere. What's the actual maximum this format can handle?

File: src/cardano_fmt_plug.c
Comment (solardiz) [inline src/cardano_fmt_plug.c:117]: We normally don't indent `#pragma` directives - just keep them on the leftmost column. Ditto for the other one below.

File: src/cardano_fmt_plug.c
Comment (solardiz) [inline src/cardano_fmt_plug.c]: Looks like the `dsks` array can be replaced with one `dsk` - only one element is used at a time. It's the same as `pk` and `ckey` in that respect - so for consistency don't use arrays for any.

`kdf_out` is different and actually needs to stay an array, because it's produced by a separate loop that can use SIMD, thus computing multiple KDF values concurrently.

File: src/ed25519-donna/ed25519-donna.c
Comment (solardiz) [inline src/ed25519-donna/ed25519-donna.c:60]: You're changing tabs to spaces here - please don't, please keep the indentation consistent with the original.

**Review by solardiz — APPROVED** [review]
This is potentially merge-able now, but I'm suggesting some more changes for you to make before we merge.

You could also want to add an entry to `doc/NEWS`.

File: src/cardano_common.h
Comment (ilap) [reply src/cardano_common.h]: I have removed all unnecessary includes.

File: src/cardano_common.h
Comment (ilap) [reply src/cardano_common.h]: Fixed

File: src/cardano_common_plug.c
Comment (ilap) [reply src/cardano_common_plug.c:60]: Fixed

File: src/cardano_common.h
Comment (ilap) [reply src/cardano_common.h]: Fixed

File: src/cardano_common_plug.c
Comment (ilap) [reply src/cardano_common_plug.c:52]: Fixed

File: src/cardano_fmt_plug.c
Comment (ilap) [reply src/cardano_fmt_plug.c:55]: Fixed

File: src/cardano_fmt_plug.c
Comment (ilap) [reply src/cardano_fmt_plug.c:117]: Fixed

File: src/cardano_fmt_plug.c
Comment (ilap) [reply src/cardano_fmt_plug.c]: Yeah, you're right, fixed.

File: src/ed25519-donna/ed25519-donna.c
Comment (ilap) [reply src/ed25519-donna/ed25519-donna.c:60]: Fixed

File: src/cardano_fmt_plug.c
Comment (ilap) [reply src/cardano_fmt_plug.c:59]: Set to 64, as it's unrealistic to have longer user's password.

File: src/cardano_fmt_plug.c
Comment (ilap) [reply src/cardano_fmt_plug.c:30]: I have removed all unnecessary ones.

File: doc/NEWS
Comment (solardiz) [inline doc/NEWS]: This should be at the end of the list, around line 256.

File: src/cardano_common.h
Comment (solardiz) [inline src/cardano_common.h]: I don't see you use anything from `#include "common.h"` in this file.

File: src/cardano_common_plug.c
Comment (solardiz) [inline src/cardano_common_plug.c:19]: I didn't mean you'd rely on includes by `cardano_common.h` to also include what's needed by files including it.

File: src/cardano_common_plug.c
Comment (solardiz) [inline src/cardano_common_plug.c]: I didn't mean you'd drop the `$1` component - I meant you'd only insist on the value being 1 in `valid` and then not parse/store it (and don't have that struct field, which you now correctly removed) until later. But OK, let's stay without the field as well.

File: src/cardano_fmt_plug.c
Comment (solardiz) [reply src/cardano_fmt_plug.c:30]: I think we differ in what we consider "unnecessary". I meant you'd remove includes of stuff that you don't use. However, you seem to have removed everything without which the program won't build. That approach results in it building "by luck" - due to includes of other needed headers from some of the headers that you still include her. I think that's not ideal.

File: src/cardano_fmt_plug.c
Comment (solardiz) [inline src/cardano_fmt_plug.c:48]: `SALT_ALIGN` can now be 1 now that you removed the `uint32_t` field from there.

File: src/cardano_fmt_plug.c
Comment (solardiz) [reply src/cardano_fmt_plug.c:59]: That's now how we do it in JtR. We determine the actual maximum (the lower of what our code supports and what the target software's code supports) and use that here unless there's performance advantage to use a smaller limit (shouldn't be the case here). Also, that `#if` and `#undef` logic should be unneeded here.

**Review by solardiz — APPROVED** [review]
I made some comments again, but for most of them I think we got to a point where it's quicker to merge this as-is and then for me or someone else to polish it a bit maybe as otherwise we'd keep running into misunderstandings on minor detail.

File: src/cardano_fmt_plug.c
Comment (solardiz) [reply src/cardano_fmt_plug.c:59]: Given that you're passing passwords through `blake2b()`, which doesn't have a low limit, our global limit of 125 should apply. So just set `PLAINTEXT_LENGTH` to 125.

File: doc/NEWS
Comment (ilap) [reply doc/NEWS]: Thx, on it.

File: src/cardano_common_plug.c
Comment (ilap) [reply src/cardano_common_plug.c]: I am reverting it  to $1 as it seems to me much clearer.

File: src/cardano_fmt_plug.c
Comment (ilap) [reply src/cardano_fmt_plug.c:48]: We would need it as I reverted the 1.

File: src/cardano_common.h
Comment (ilap) [reply src/cardano_common.h]: fixing it

File: src/cardano_common_plug.c
Comment (ilap) [inline src/cardano_common_plug.c:51]: Ooops, fixing this.

**Comment by solardiz** [comment]
I'll merge this now, and will possibly make further adjustments in a separate commit of my own. Thank you!

File: src/cardano_fmt_plug.c
Comment (solardiz) [reply src/cardano_fmt_plug.c:48]: You didn't reintroduce the struct field, so we can have `SALT_ALIGN` at 1.

**Pull request merged**
